### ISSUE TITLE
Document DYLD_FALLBACK_LIBRARY_PATH behavior from Big Sur on

### ISSUE THREAD

**Issue by sethrj** [spack/spack#24306]
Spack-generated lmod files are defining DYLD_FALLBACK_LIBRARY_PATH as a shell variable but (silently) not exporting it. This is apparently a new  security feature on Big Sur. I'm not 100%  sure I understand the behavior yet: whether the issue only shows up with forking, with the `env` command, or with bash. See [this stackexchange question](https://apple.stackexchange.com/questions/212945/unable-to-set-dyld-fallback-library-path-in-shell-on-osx-10-11-1) and [this Apple documentation on System Integrity Protection](https://developer.apple.com/library/archive/documentation/Security/Conceptual/System_Integrity_Protection_Guide/RuntimeProtections/RuntimeProtections.html).

We might want/have to  remove `DYLD_FALLBACK_LIBRARY_PATH` or add some special exception? because it's *very* confusing behavior for a user to do `export VAR=VALUE` and see `echo $VAR` work but then have `env | grep VAR` hide it. I only asked the user to do this out of desperation (eliminating the impossible, etc.).

@adamjstewart I know you're a mac user so maybe you've seen this? I only noticed it because I was using `DYLD_FALLBACK_LIBRARY_PATH` to internally set up the `CMAKE_INSTALL_RPATH` on an in-house code.

### Steps to reproduce the issue

```console
$ echo $BASH_VERSION
5.1.8(1)-release
$ type module
module is a function
module ()
{
    eval $($LMOD_CMD bash "$@") && eval $(${LMOD_SETTARG_CMD:-:} -s sh)
}
$ $LMOD_CMD bash load zlib > loadme.sh; chmod a+x loadme.sh
$ grep DYLD loadme.sh
DYLD_FALLBACK_LIBRARY_PATH=/rnsdhpc/code/spack/opt/spack/apple-clang/zlib/w2ka2t3/lib;
export DYLD_FALLBACK_LIBRARY_PATH;
$ source loadme.sh
$ echo $DYLD_FALLBACK_LIBRARY_PATH
/rnsdhpc/code/spack/opt/spack/apple-clang/zlib/w2ka2t3/lib
$ env | grep DYLD
__LMOD_REF_COUNT_DYLD_FALLBACK_LIBRARY_PATH=/rnsdhpc/code/spack/opt/spack/apple-clang/zlib/w2ka2t3/lib:1
$ echo $PKG_CONFIG_PATH
/rnsdhpc/code/spack/opt/spack/apple-clang/zlib/w2ka2t3/lib/pkgconfig
$ env | grep PKG_CONFIG_PATH
PKG_CONFIG_PATH=/rnsdhpc/code/spack/opt/spack/apple-clang/zlib/w2ka2t3/lib/pkgconfig
__LMOD_REF_COUNT_PKG_CONFIG_PATH=/rnsdhpc/code/spack/opt/spack/apple-clang/zlib/w2ka2t3/lib/pkgconfig:1
```

### Information on your system

* **Spack:** 0.16.2-3119-2b53d84a56
* **Python:** 3.8.5
* **Platform:** darwin-bigsur-skylake
* **Concretizer:** original

**Comment by adamjstewart** [comment]
In my experience, `DYLD_LIBRARY_PATH` and `DYLD_FALLBACK_LIBRARY_PATH` are unset/ignored whenever you run any system executable. As you've noted, this is a "feature" of SIP. Notably, if you have a script with a shebang like `/usr/bin/env python`, if you run it like `./script.py` it won't work, but if you run it like `python script.py` it will. You could disable SIP, but I don't recommend this.

**Comment by sethrj** [comment]
Right, that makes sense, and I def don't want to disable SIP. However it is surprising to run CMake (and that installed through homebrew) after loading the spack module and find that `DYLD_FALLBACK_LIBRARY_PATH` isn't accessible to a forked process.

Are the DYLD paths generally needed  for spack-installed software?  I guess I could at least add a note to the spack documentation about this system behavior.

**Comment by adamjstewart** [comment]
I've found them necessary mostly for Python packages where RPATH isn't always supported.

**Comment by sethrj** [comment]
OK, I'll just add a documentation note about the unusual behavior of the DYLD guys to spack's `modules` documentation.

**Comment by 06alesso** [comment]
Hi! I'd like to work on this issue and add documentation about the DYLD_FALLBACK_LIBRARY_PATH behavior on macOS Big Sur and later. Is this still available to work on?

**Comment by sethrj** [comment]
I've learned part of the problem since I opened this issue five years ago: you don't see the filtered environment in  `env | grep` because `env` is a command (and so DYLD_ is filtered out before passing it into it). If you just `echo $DYLD_…` (using a shell builtin) it shows up.
